### ISSUE TITLE
Potential misemulation of upper bits in CGA 2bpp compatibility mode on VGA

### ISSUE THREAD

**Issue by iamgreaser** [86Box/86Box#3836]
### What happened?

I have a strong suspicion that the 2bpp mode is emulated incorrectly on VGA.
Specifically, this is controlled by bits 5,6 of Graphics Controller register 0x05 being set to 01.

I believe that the correct behaviour for the 2bpp mode is such that, not only does the lower 2 planes get rearranged to a chunky rearrangement such that plane 0 is on the left and plane 1 is on the right, but the same happens with planes 2 and 3 as well, effectively doing some hybrid 2x2 chunky x planar mode.

The way that 86Box currently renders this mode is to completely ignore planes 2 and 3. However, the correct way to mask those planes out is via the Attribute Controller.

The documentation for the 82C451 VGA clone chipset explicitly spells out the behaviour performed there, while the OTI-037 documentation and the official IBM documentation state that it applies to *both* pairs of planes in a rather non-obvious way.

IBM (and almost the same for OTI-037):

> When set to 1, the Shift Register Mode field (bit 5) directs the shift registers in the graphics controller to format the serial data stream with even-numbered bits **from both maps** on even-numbered maps, and odd-numbered bits **from both maps** on the odd-numbered maps. This bit is used for modes 4 and 5.

From the 82C451 documentation:

![image](https://github.com/86Box/86Box/assets/138113/bcee7252-4002-42d0-98ea-44aa208a8d55)

(Side note: The "Note" at the bottom is impossible to achieve on an EGA due to the Graphics Controllers being... well, 2 chips instead of 1, and they aren't connected up to allow that. If that's how VGA behaves then that could be a cheeky way to make something appear differently on VGA vs. EGA...)

For comparison:
- DOSBox-X agrees with my interpretation.
- The EGA emulation also agrees with my interpretation, although that's partly because I was the one who rewrote the scanline renderer, so there may be some confirmation bias.
- DOSBox 0.74-3 is completely wrong, as it appears to completely ignore the plane write selection register.
- QEMU seems to be treating what I've set up as being a planar mode.

Attached here is a VM and a bootable test program (assemble the disk image with FASM 1):

[vga2bpp-86boxvm.zip](https://github.com/86Box/86Box/files/13407150/vga2bpp-86boxvm.zip)

I believe this is the correct interpretation, and this is what the EGA renderer (as well as the SuperEGA renderer) produces:

![image](https://github.com/86Box/86Box/assets/138113/ac4e8e57-d5ad-4e74-bbdc-47d3ec167443)

This is what the VGA renderer produces, and at least one SVGA renderer produces:

![image](https://github.com/86Box/86Box/assets/138113/f29ba08c-b6f7-41e8-8a87-4998080ffdc2)

### Configuration file

```ini
[General]
vid_renderer = qt_software
dpi_scale = 0

[Machine]
machine = mr286
cpu_family = 286
cpu_speed = 25000000
cpu_multi = 1
cpu_use_dynarec = 0
fpu_softfloat = 0
time_sync = utc
cpu_waitstates = 1
cpu = 6
mem_size = 512

[Video]
gfxcard = vga

[Input devices]
mouse_type = none

[Sound]
fm_driver = nuked

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = none
cassette_mode = load

[Floppy and CD-ROM drives]
fdd_01_fn = 2x2tst0b.img
fdd_01_writeprot = 1

[SuperEGA]
memory = 256
monitor_type = 9
```

### Operating system

Artix Linux

### CPU

AMD Ryzen 9 7950X3D 16-Core Processor

### 86Box version

Commit at 2023-11-12T23:55:00-0500, hash cf8dea5ed3c0dac1aea6ab169ebc5711af559fde

### Build architecture

Linux - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

I built 86Box myself (please tell us more about your build configuration)

### Additional context

I can be found on the RoL 86Box IRC as GreaseMonkey. I am happy to fix the 2bpp renderers (ideally by making them go via the 4bpp renderers with some remapping tables), but the prospect of doing something as extensive as the EGA renderer rewrite scares me from a compatibility perspective as well as a performance perspective - AFAIK, this affects every SVGA-capable card, too.

**Comment by iamgreaser** [comment]
An update with some findings.

I performed some simpler tests on a GM45 laptop from 2009 with an Intel GMA 4500MHD which agrees with this interpretation, using a simpler test from `debug` (specifically using FreeDOS running off a bootable CD I put together some years ago):

```
a100
mov ax,0010
int 10
mov dx,03ce
mov ax,2005
out dx,ax
int 3

g=100
```

According to my interpretation, if the text colour is 0x07, then one should see each character duplicated in half its usual width, with the first character mostly white and the second character mostly dark cyan. If the upper 2 planes are not used, then one will see both duplicates as dark cyan.

Also it's likely that the moment one has enough text to shove onto the screen to make it scroll, the mode will reset to non-chunky.

Unfortunately, I don't have the original hardware. Maybe one day someone will. For now, once I can find a spare VGA monitor the best thing I'll actually be able to do is an S3 Trident, which is an onboard S3 Trio of some sort.

Anyway, it's looking like the right thing to do is to at the very least get the 2bpp modes fixed.

**Comment by iamgreaser** [comment]
From one misemulation to another... I have merged a bunch of renderers together in my gm/vga-renderer-merge-and-fixup-3836 branch, and I'm starting to understand what the `force_old_addr` hack is for, so it works fine on IBM VGA as well as an S3 Trio...
But now I've got VGA looking like this using the test program/floppy bootloader thing:

![image](https://github.com/86Box/86Box/assets/138113/6a6b1ad3-2cd2-49bb-ac90-e23883141ed3)

What's happening here is that, well... I did something wrong in the test program. We're dealing with 8 4bpp pixels here, which means I should be incrementing the CRTC address counter on every character clock, but I set it up to increment every two clocks without thinking it through. (The other flag, which shifts the address left by 1, was correctly set.)

This also reveals a potential misemulation in the EGA renderer, which seems to be ignoring this flag. Due to there being a difference between "load every n clocks" (Sequencer) and "increase address counter every n clocks" (CRTC) which I don't think I'm emulating in the EGA renderer at the moment.

I'll need to test some stuff on my GM45 laptop as a sanity check.

With that said, the bug mentioned in this issue *is* fixed, but I've very much crashed into the iceberg at this point, so I'll keep meditating on this.
